crates/polars-plan/src/plans/ir/mod.rs

```diff
@@ -203,6 +210,844 @@ impl IRPlan {
+        let mut data_map = PlHashMap::new();
+
+        for (id, df) in dfs.iter().enumerate() {
+            let schema = df.schema().clone();
+            let data_node = data_arena.add(IR::DataFrameScan {
+                df: df.clone(),
+                schema,
+                output_schema: None,
+            });
+            data_map.insert(id, data_node);
+        }
+
+        self.bind_data(data_map, &data_arena)
+    }
+
+    /// Count the number of PlaceholderScan nodes in the IR plan.
+    ///
+    /// Placeholders are encountered and assigned IDs (0, 1, 2, ...) during a depth-first
+    /// traversal of the IR tree in `to_template()`. This same traversal order determines
+    /// the mapping between DataFrame positions in `bind_to_dfs()` and placeholder IDs.
+    fn count_placeholders(&self) -> usize {
+        Self::count_placeholders_recursive(self.lp_top, &self.lp_arena)
+    }
+
+    /// Recursively count PlaceholderScan nodes in the IR tree.
+    ///
+    /// This traversal follows the same depth-first order used when assigning placeholder IDs,
+    /// ensuring consistent ordering across template creation and data binding operations.
+    fn count_placeholders_recursive(node: Node, arena: &Arena<IR>) -> usize {
+        let ir = arena.get(node);
+        match ir {
+            IR::PlaceholderScan { .. } => 1,
+            _ => {
+                // Sum placeholder counts from all input nodes
+                ir.inputs()
+                    .map(|input| Self::count_placeholders_recursive(input, arena))
+                    .sum()
+            },
+        }
+    }
+
+    #[recursive::recursive]
+    fn replace_placeholder(
+        node: Node,
+        data_map: &PlHashMap<usize, Node>,
+        data_arena: &Arena<IR>,
+        template_arena: &Arena<IR>,
+        new_arena: &mut Arena<IR>,
+    ) -> PolarsResult<Node> {
+        let ir = template_arena.get(node);
+        let new_ir = match ir {
+            IR::PlaceholderScan { id, schema, .. } => {
+                let data_node = data_map.get(id).ok_or_else(
+                    || polars_err!(ComputeError: "Placeholder ID {} not found in data map", id),
+                )?;
+
+                let data_ir = data_arena.get(*data_node);
+                let data_schema = match data_ir {
+                    IR::DataFrameScan {
+                        schema: data_schema,
+                        ..
+                    } => data_schema,
+                    IR::Scan { file_info, .. } => &file_info.schema,
+                    #[cfg(feature = "python")]
+                    IR::PythonScan { options } => &options.schema,
+                    _ => polars_bail!(ComputeError:
+                        "bind_data requires data to be a data-source node (DataFrameScan, Scan, or PythonScan)"
+                    ),
+                };
+
+                // Allow empty schemas to bind to any data (generic templates)
+                if !schema.is_empty() {
```

Comment:
Maybe this could be reused: https://github.com/pola-rs/polars/blob/rs-0.52.0/crates/polars-core/src/schema/mod.rs#L139
(or did you want to inspect manually to customize the error messages?)